core/src/main/scala/org/apache/spark/h2o/converters/WriteConverterCtxUtils.scala

```diff
@@ -56,69 +56,139 @@ object WriteConverterCtxUtils {
-      // this is only required in internal cluster mode
-      val prefs = hc.getH2ONodes().map { nodeDesc =>
-        s"executor_${nodeDesc.hostname}_${nodeDesc.nodeId}"
+      writerClient.initFrame(keyName, colNames)
+
+
+      val rdd = if (hc.getConf.runsInInternalClusterMode) {
+        // this is only required in internal cluster mode
+        val prefs = hc.getH2ONodes().map { nodeDesc =>
+          s"executor_${nodeDesc.hostname}_${nodeDesc.nodeId}"
+        }
+        new H2OAwareRDD(prefs, rddInput)
+      } else {
+        rddInput
       }
-      new H2OAwareRDD(prefs, rddInput)
-    } else {
-      rddInput
-    }
 
-    val nonEmptyPartitions = rdd.mapPartitionsWithIndex {
-      case (idx, it) => if (it.nonEmpty) Iterator.single(idx) else Iterator.empty
-    }.collect().toSeq.sorted
+      val nonEmptyPartitions = getNonEmptyPartitions(rdd)
 
-    // prepare required metadata based on the used backend
-    val uploadPlan = if (hc.getConf.runsInExternalClusterMode) {
-      Some(ExternalWriteConverterCtx.scheduleUpload(nonEmptyPartitions.size))
-    } else {
-      None
-    }
-    val operation: SparkJob[T] = func(keyName, expectedTypes, uploadPlan, writeTimeout, H2O.SELF.getTimestamp(), sparse, nonEmptyPartitions)
-    val rows = hc.sparkContext.runJob(rdd, operation, nonEmptyPartitions) // eager, not lazy, evaluation
-    val res = new Array[Long](nonEmptyPartitions.size)
-    rows.foreach { case (cidx, nrows) => res(cidx) = nrows }
-    // Add Vec headers per-Chunk, and finalize the H2O Frame
-
-    // get the vector types from expected types in case of external h2o cluster
-    val types = if (hc.getConf.runsInExternalClusterMode) {
-      ExternalFrameUtils.vecTypesFromExpectedTypes(expectedTypes, maxVecSizes)
-    } else {
-      expectedTypes
+      // prepare required metadata based on the used backend
+      val uploadPlan = if (hc.getConf.runsInExternalClusterMode) {
+        Some(ExternalWriteConverterCtx.scheduleUpload(nonEmptyPartitions.size))
+      } else {
+        None
+      }
+      val operation: SparkJob[T] = func(keyName, expectedTypes, uploadPlan, writeTimeout, H2O.SELF.getTimestamp(), sparse, nonEmptyPartitions)
+      val rows = hc.sparkContext.runJob(rdd, operation, nonEmptyPartitions) // eager, not lazy, evaluation
+      val res = new Array[Long](nonEmptyPartitions.size)
+      rows.foreach { case (cidx, nrows) => res(cidx) = nrows }
+      // Add Vec headers per-Chunk, and finalize the H2O Frame
+
+      // get the vector types from expected types in case of external h2o cluster
+      val types = if (hc.getConf.runsInExternalClusterMode) {
+        ExternalFrameUtils.vecTypesFromExpectedTypes(expectedTypes, maxVecSizes)
+      } else {
+        expectedTypes
+      }
+
+      writerClient.finalizeFrame(keyName, res, types)
+      keyName
     }
 
-   writerClient.finalizeFrame(keyName, res, types)
   }
 
+  object RESTBasedConverter extends Converter {
```

Comment:
I would rename this to `RestBasedExternalConvertor`, because:

I think that we at some point will get rid of the client, but I would still distinguish between internal & external rest API convertors. The external sends the data, the internal will just notify the right executor/h2o node via rest api that it should convert one partition to H2O chunk. But that is for the future.

What do you think?

Reply:
No problem, will do ;-)